sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala

```diff
@@ -65,55 +67,97 @@ private[hive] class HiveFunctionRegistry(
 
       val functionClassName = functionInfo.getFunctionClass.getName
 
-      // When we instantiate hive UDF wrapper class, we may throw exception if the input expressions
-      // don't satisfy the hive UDF, such as type mismatch, input number mismatch, etc. Here we
-      // catch the exception and throw AnalysisException instead.
+      // Make an expression of the appropriate type based on the function class.
       try {
         if (classOf[GenericUDFMacro].isAssignableFrom(functionInfo.getFunctionClass)) {
           val udf = HiveGenericUDF(
             name, new HiveFunctionWrapper(functionClassName, functionInfo.getGenericUDF), children)
           udf.dataType // Force it to check input data types.
           udf
-        } else if (classOf[UDF].isAssignableFrom(functionInfo.getFunctionClass)) {
-          val udf = HiveSimpleUDF(name, new HiveFunctionWrapper(functionClassName), children)
+        } else {
+          makeFunctionBuilder(name, functionInfo.getFunctionClass)(children)
+        }
+      } catch {
+        case ae: AnalysisException =>
+          throw ae
+        case NonFatal(e) =>
+          // TODO: don't swallow stack trace here
+          throw new AnalysisException(
+            s"No handler for Hive UDF '${functionInfo.getFunctionClass.getCanonicalName}': $e")
+      }
+    }
+  }
+
+  /**
+   * Construct a [[FunctionBuilder]] based on the provided class that represents a function.
+   *
+   * This performs reflection to decide what type of [[Expression]] to return in the builder.
+   * This is useful for creating temporary functions.
+   */
+  override def makeFunctionBuilder(funcName: String, funcClassName: String): FunctionBuilder = {
+    makeFunctionBuilder(funcName, Utils.classForName(funcClassName))
+  }
+
+  /**
+   * Construct a [[FunctionBuilder]] based on the provided class that represents a function.
+   */
+  private def makeFunctionBuilder(name: String, clazz: Class[_]): FunctionBuilder = {
+    // When we instantiate hive UDF wrapper class, we may throw exception if the input
+    // expressions don't satisfy the hive UDF, such as type mismatch, input number
+    // mismatch, etc. Here we catch the exception and throw AnalysisException instead.
+    try {
+      if (classOf[UDF].isAssignableFrom(clazz)) {
+        (children: Seq[Expression]) => {
+          val udf = HiveSimpleUDF(name, new HiveFunctionWrapper(clazz.getName), children)
           udf.dataType // Force it to check input data types.
           udf
```

Comment:
Looks like we want to do the try/catch inside the function because we will not trigger udf.dataType when we create the builder at here? Like

```
(children: Seq[Expression]) => {
  try {
    ...
  }  catch {
    ...
  }
}
```